tests/Integration/Behaviour/Features/Context/Domain/Product/ProductCombinationFeatureContext.php

```diff
@@ -0,0 +1,116 @@
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace Tests\Integration\Behaviour\Features\Context\Domain\Product;
+
+use Behat\Gherkin\Node\TableNode;
+use PHPUnit\Framework\Assert;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\Command\GenerateProductCombinationsCommand;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\Query\GetProductCombinationsForEditing;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\QueryResult\CombinationForEditing;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\QueryResult\CombinationListForEditing;
+use PrestaShop\PrestaShop\Core\Domain\Product\Exception\ProductException;
+use Tests\Integration\Behaviour\Features\Context\Util\PrimitiveUtils;
+
+class ProductCombinationFeatureContext extends AbstractProductFeatureContext
+{
+    /**
+     * @When I generate combinations for product :productReference using following attributes:
+     *
+     * @param string $productReference
+     * @param TableNode $table
+     */
+    public function generateCombinations(string $productReference, TableNode $table): void
+    {
+        $tableData = $table->getRowsHash();
+        $groupedAttributeIds = $this->parseGroupedAttributeIds($tableData);
+
+        try {
+            $this->getCommandBus()->handle(new GenerateProductCombinationsCommand(
+                $this->getSharedStorage()->get($productReference),
+                $groupedAttributeIds
+            ));
+        } catch (ProductException $e) {
+            $this->setLastException($e);
+        }
+    }
+
+    /**
+     * @Then product :productReference should have following combinations:
+     *
+     * @param string $productReference
+     * @param TableNode $table
+     */
+    public function assertProductCombinations(string $productReference, TableNode $table): void
+    {
+        $dataRows = $table->getColumnsHash();
+        $expectedCombinationsCount = count($dataRows);
+
+        /** @var CombinationListForEditing $combinationsForEditing */
+        $combinationsForEditing = $this->getQueryBus()->handle(new GetProductCombinationsForEditing(
+            $this->getSharedStorage()->get($productReference),
+            $this->getDefaultLangId(),
+            $expectedCombinationsCount
+        ));
+
+        Assert::assertEquals(
+            count($combinationsForEditing->getCombinations()),
+            $expectedCombinationsCount,
+            'Unexpected combinations count'
+        );
+
+        /** @var CombinationForEditing $combinationForEditing */
+        foreach ($combinationsForEditing as $key => $combinationForEditing) {
```

Comment:
```suggestion
        foreach ($combinationsForEditing->getCombinations() as $key => $combinationForEditing) {
```

Reply:
not sure how tests are working when there is a critical error like this. I guess the coverage is way too low

Reply:
Well the object is not Traversable so I guess the loop was simply ignored 🤷‍♂️